src/models/transaction.ts

```diff
@@ -0,0 +1,350 @@
+
+    // Len tokens
+    if (this.tokens.length) {
+      // Create token tx does not have tokens array
+      arr.push(helpers.intToBytes(this.tokens.length, 1))
+    }
+
+    // Len inputs
+    arr.push(helpers.intToBytes(this.inputs.length, 1))
+    // Len outputs
+    arr.push(helpers.intToBytes(this.outputs.length, 1))
+
+    // Tokens data
+    for (const token of this.tokens) {
+      arr.push(new encoding.BufferReader(token).buf);
+    }
+
+    for (const inputTx of this.inputs) {
+      arr.push(util.buffer.hexToBuffer(inputTx.hash));
+      arr.push(helpers.intToBytes(inputTx.index, 1));
+      // Input data will be fixed to 0 for now
+      arr.push(helpers.intToBytes(0, 2));
+    }
+
+    for (const outputTx of this.outputs) {
+      arr.push(outputTx.valueToBytes());
+      // Token data
+      arr.push(helpers.intToBytes(outputTx.tokenData, 1));
+
+      const outputScript = outputTx.createScript();
+      arr.push(helpers.intToBytes(outputScript.length, 2));
+      arr.push(outputScript);
+    }
+
+    if (this.version === CREATE_TOKEN_TX_VERSION) {
+      // Create token tx need to add extra information
+      arr = [...arr, ...this.serializeTokenInfo()];
+    }
+
+    return util.buffer.concat(arr);
+  }
+
+  /*
+   * Execute hash of the data to sign
+   *
+   * @return {Buffer} data to sign hashed
+   *
+   * @memberof Transaction
+   * @inner
+   */
+  getDataToSignHash(): Buffer {
+    const dataToSign = this.getDataToSign();
+    const hashbuf = crypto.Hash.sha256sha256(dataToSign);
+    return new encoding.BufferReader(hashbuf).readReverse();
+  }
+
+  /**
+   * Calculate the minimum tx weight
+   * 
+   * @throws {ConstantNotSet} If the weight constants are not set yet
+   *
+   * @return {number} Minimum weight calculated (float)
+   * @memberof Transaction
+   * @inner
+   */
+  calculateWeight(): number {
+    let txSize = this.toBytes().length;
+
+    // If parents are not in txData, we need to consider them here
+    if (!this.parents || !this.parents.length || this.parents.length === 0) {
+      // Parents are always two and have 32 bytes each
+      txSize += 64;
+    }
+
+    let sumOutputs = this.getOutputsSum();
+    // Preventing division by 0 when handling authority methods that have no outputs
+    sumOutputs = Math.max(1, sumOutputs);
+
+    // We need to take into consideration the decimal places because it is inside the amount.
+    // For instance, if one wants to transfer 20 HTRs, the amount will be 2000.
+    const amount = sumOutputs / (10 ** DECIMAL_PLACES);
+
+    let weight = (TX_WEIGHT_CONSTANTS.txWeightCoefficient * Math.log2(txSize) + 4 / (1 + TX_WEIGHT_CONSTANTS.txMinWeightK / amount) + 4);
+
+    // Make sure the calculated weight is at least the minimum
+    weight = Math.max(weight, TX_WEIGHT_CONSTANTS.txMinWeight)
+    // FIXME precision difference between backend and frontend (weight (17.76246721531992) is smaller than the minimum weight (17.762467215319923))
+    return weight + 1e-6;
```

Comment:
We can explain that `1e-6` have no practical effect when mining the transaction.

Reply:
Done